sdks/java/harness/src/main/java/org/apache/beam/fn/harness/SplittablePairWithRestrictionDoFnRunner.java

```diff
@@ -0,0 +1,384 @@
+    @SuppressWarnings("rawtypes") // passed to FnApiStateAccessor which uses rawtypes
+    ImmutableMap.Builder<TupleTag<?>, SideInputSpec> tagToSideInputSpecMapBuilder =
+        ImmutableMap.builder();
+    for (Map.Entry<String, RunnerApi.SideInput> entry :
+        parDoPayload.getSideInputsMap().entrySet()) {
+      String sideInputTag = entry.getKey();
+      RunnerApi.SideInput sideInput = entry.getValue();
+      PCollection sideInputPCollection =
+          components.getPcollectionsMap().get(pTransform.getInputsOrThrow(sideInputTag));
+      WindowingStrategy<?, ?> sideInputWindowingStrategy =
+          rehydratedComponents.getWindowingStrategy(sideInputPCollection.getWindowingStrategyId());
+      tagToSideInputSpecMapBuilder.put(
+          new TupleTag<>(entry.getKey()),
+          SideInputSpec.create(
+              sideInput.getAccessPattern().getUrn(),
+              rehydratedComponents.getCoder(sideInputPCollection.getCoderId()),
+              (Coder<WindowT>) sideInputWindowingStrategy.getWindowFn().windowCoder(),
+              PCollectionViewTranslation.viewFnFromProto(entry.getValue().getViewFn()),
+              (WindowMappingFn<WindowT>)
+                  PCollectionViewTranslation.windowMappingFnFromProto(
+                      entry.getValue().getWindowMappingFn())));
+    }
+    @SuppressWarnings("rawtypes") // passed to FnApiStateAccessor which uses rawtypes
+    Map<TupleTag<?>, SideInputSpec> tagToSideInputSpecMap = tagToSideInputSpecMapBuilder.build();
+
+    // Register processing methods
+    if (doFnSignature.getInitialRestriction().observesWindow() || !sideInputMapping.isEmpty()) {
+      addPCollectionConsumer.accept(
+          mainInputName, this::processElementForWindowObservingPairWithRestriction);
+      this.mutableArgumentProvider = new WindowObservingProcessBundleContext();
+    } else {
+      addPCollectionConsumer.accept(mainInputName, this::processElementForPairWithRestriction);
+      this.mutableArgumentProvider = new NonWindowObservingProcessBundleContext();
+    }
+    addTearDownFunction.accept(this::tearDown);
+
+    // State accessor for supporting side inputs, requires window coder and lots of metadata
+    WindowingStrategy<?, ?> windowingStrategy =
+        rehydratedComponents.getWindowingStrategy(mainInput.getWindowingStrategyId());
+    Coder<WindowT> windowCoder = (Coder<WindowT>) windowingStrategy.getWindowFn().windowCoder();
+    this.stateAccessor =
+        new FnApiStateAccessor<>(
+            pipelineOptions,
+            runnerCapabilities,
+            pTransformId,
+            processBundleInstructionId,
+            cacheTokens,
+            bundleCache,
+            processWideCache,
+            tagToSideInputSpecMap,
+            beamFnStateClient,
+            null,
+            (Coder<BoundedWindow>) windowCoder,
+            null,
+            mutableArgumentProvider::getCurrentWindow);
```

Comment:
I would actually like to move the side input and state accessor logic out of all the classes and share it, but for now it is tied to their inner classes methods. I want to take this refactor a step at a time. Eventually we should have one runner per `processElementForXYZ` basically, and then share some initialization code.

Reply:
Done in the next PR